README.md
# chi2_tool
Tool for computing chi2 maps from .ROOT files


CHANGELOG.rst
Unreleased changes
------------------

* Remove template artifacts


Version 0
---------

0.1.0 / 2023-03-24
~~~~~~~~~~~~~~~~~~
* Project generated using the cookiecutter template from
  https://git.km3net.de/templates/python-project


test_functions.py
import numpy as np
import uproot
import matplotlib.pyplot as plt
import matplotlib as mpl
import boost_histogram as bh
from functions import *
import lnl


f = uproot.open(
    "/sps/km3net/users/alflazo/dstProd/v7.1_v7.2_jsh/mcv7.2.gsg_muon-CC_1-100GeV.km3sim.jorcarec.jsh.aanet.dst_merged.root"
)  # nu mu CC
# f = uproot.open("/sps/km3net/users/alflazo/dstProd/v7.1_v7.2_jsh/mcv7.2.gsg_elec-CC_1-100GeV.km3sim.jorcarec.jsh.aanet.dst_merged.root") # nu e CC
# f = uproot.open("/sps/km3net/users/alflazo/dstProd/v7.1_v7.2_jsh/mcv7.2.gsg_muon-NC_1-100GeV.km3sim.jorcarec.jsh.aanet.dst_merged.root") # nu mu NC
# f = uproot.open("/sps/km3net/users/spenamar/oscillations/chi2_tool/results/v7.10/SelectedEvents_mc_nu.root") # merged DST of events passing selection
# E = f['E:Evt/mc_trks/mc_trks.E'].array()[:,0].to_numpy()
# dir_z = f['E:Evt/mc_trks/mc_trks.dir.z'].array()[:,0].to_numpy()
E = f["E:Evt/trks/trks.E"].array()[:, 0].to_numpy()
dir_z = f["E:Evt/trks/trks.dir.z"].array()[:, 0].to_numpy()
nu_type = f["E:Evt/mc_trks/mc_trks.type"].array()[:, 0].to_numpy()
is_cc = f["T:sum_mc_nu/cc"].array().to_numpy()
print(is_cc)
w2 = f["E:Evt/w"].array()[:, 1].to_numpy()
w_noOsc = f["T:sum_mc_evt/weight_noOsc"].array().to_numpy()
w_Osc = f["T:sum_mc_evt/weight"].array().to_numpy()
livetime = f["T:sum_mc_evt/livetime_DAQ"].array().to_numpy()
ngen = f["T:sum_mc_evt/n_gen"].array().to_numpy()


osci_w = compute_evt_weight(nu_type, E, dir_z, is_cc, w2, livetime, ngen)
nu_params = {
    "dm_21": 7.42e-5,
    "dm_31": 2.210e-3,
    # "dm_31": 0,
    "theta_12": 40.0 * np.pi / 180,
    "theta_23": 70.0 * np.pi / 180,
    # "theta_23": 0. * np.pi/180,
    "theta_13": 8.62 * np.pi / 180,
    "dcp": 230 * np.pi / 180,
}
osci_w_bad = compute_evt_weight(
    nu_type, E, dir_z, is_cc, w2, livetime, ngen, nu_params=nu_params
)
print(osci_w)
print(osci_w_bad)


# plothist = lambda h: plt.bar(*h.axes.centers, h.values(), width=h.axes.widths[0], alpha=0.6);

h1 = bh.Histogram(
    bh.axis.Regular(50, 1, 100, transform=bh.axis.transform.log),
)
h2 = bh.Histogram(
    bh.axis.Regular(50, 1, 100, transform=bh.axis.transform.log),
)
h3 = bh.Histogram(
    bh.axis.Regular(50, 1, 100, transform=bh.axis.transform.log),
)

h1.fill(E, weight=osci_w)
h2.fill(E, weight=osci_w_bad)
fig, ax = plt.subplots()
plothist(h1, ax)
plothist(h2, ax)
plt.xscale("log")
plt.show()

h1 = bh.Histogram(
    bh.axis.Regular(50, 1, 100, transform=bh.axis.transform.log),
)
h2 = bh.Histogram(
    bh.axis.Regular(50, 1, 100, transform=bh.axis.transform.log),
)
h3 = bh.Histogram(
    bh.axis.Regular(50, 1, 100, transform=bh.axis.transform.log),
)
h1.fill(E, weight=osci_w)
h2.fill(E, weight=w_noOsc)
h3.fill(E, weight=w_Osc)
fig, ax = plt.subplots()
plothist(h1, ax)
# plothist(h2)
plothist(h3, ax)
plt.xscale("log")
plt.show()

h1 = bh.Histogram(
    bh.axis.Regular(30, 1, 100, transform=bh.axis.transform.log),
)
h2 = bh.Histogram(
    bh.axis.Regular(30, 1, 100, transform=bh.axis.transform.log),
)
h11 = bh.Histogram(
    bh.axis.Regular(30, 1, 100, transform=bh.axis.transform.log),
)
h22 = bh.Histogram(
    bh.axis.Regular(30, 1, 100, transform=bh.axis.transform.log),
)
h1.fill(E, weight=osci_w)
h2.fill(E, weight=w_noOsc)
h11.fill(E, weight=osci_w**2)
h22.fill(E, weight=w_noOsc**2)
fig, ax = plt.subplots()
plothist(h1, ax)
plothist(h2, ax)
plt.xscale("log")
plt.show()
fig, ax = plt.subplots()
plothist(h11, ax)
plothist(h22, ax)
plt.xscale("log")
plt.show()

# def plothist2d(h):
#        return plt.pcolormesh(*h.axes.edges.T, h.values().T)


Ebins = 30
zbins = 20

h1 = bh.Histogram(
    bh.axis.Regular(Ebins, 1, 100, transform=bh.axis.transform.log),
    bh.axis.Regular(zbins, -1, 0),
)
h2 = bh.Histogram(
    bh.axis.Regular(Ebins, 1, 100, transform=bh.axis.transform.log),
    bh.axis.Regular(zbins, -1, 0),
)
h11 = bh.Histogram(
    bh.axis.Regular(Ebins, 1, 100, transform=bh.axis.transform.log),
    bh.axis.Regular(zbins, -1, 0),
)
h22 = bh.Histogram(
    bh.axis.Regular(Ebins, 1, 100, transform=bh.axis.transform.log),
    bh.axis.Regular(zbins, -1, 0),
)
h1.fill(E, -1.0 * dir_z, weight=osci_w)
h2.fill(E, -1.0 * dir_z, weight=osci_w_bad)
h11.fill(E, -1.0 * dir_z, weight=osci_w**2)
h22.fill(E, -1.0 * dir_z, weight=osci_w_bad**2)
s = np.sqrt(h22.values()) / h2.values()
fig, ax = plt.subplots()
plothist2d(h1, ax)
# plothist(h2)
plt.xscale("log")
plt.show()
fig, ax = plt.subplots()
plothist2d(h2, ax)
plt.xscale("log")
plt.show()


print("#" * 20)
print("#" * 20)
print("#" * 20)
print("#" * 20)
vLnL = np.vectorize(lnl.LnL)
print("h1")
print(h1.values())
print(np.shape(h1.values()))
print("h2")
print(h2.values())
print(np.shape(h2.values()))
print("s")
print(s)
print(np.shape(s))
chi2 = vLnL(h1.values(), h2.values(), s)
print("chi2")
print(chi2)
print(np.shape(chi2))
print("#" * 20)
print("#" * 20)
print("#" * 20)
print("#" * 20)

fig, ax = plt.subplots()
h = compute_chi2_map(E, dir_z, osci_w, osci_w_bad)

plothist2d(h, ax)
plt.xscale("log")
plt.show()


tests/test_calc.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import unittest

from chi2_tool import Calculator


class TestCalculator(unittest.TestCase):
    def setUp(self):
        self.calculator = Calculator()

    def test_add(self):
        assert 1 == self.calculator.add(0, 1)
        assert 3 == self.calculator.add(1, 2)

    def test_add_negative_numbers(self):
        assert -1 == self.calculator.add(0, -1)
        assert -1 == self.calculator.add(1, -2)
        assert -3 == self.calculator.add(-1, -2)

    def test_divide(self):
        assert 0 == self.calculator.divide(0, 1)
        self.assertAlmostEqual(0.5, self.calculator.divide(1, 2))

    def test_divide_by_zero_raises(self):
        with self.assertRaises(ValueError):
            self.calculator.divide(1, 0)

    def test_multiply(self):
        assert 2 == self.calculator.multiply(1, 2)
        assert 6 == self.calculator.multiply(2, 3)

    def test_multiply_with_zero(self):
        for b in range(100):
            assert 0 == self.calculator.multiply(0, b)


LnL.cpp
#include <cmath>
//#include <iostream>
#include <pybind11/pybind11.h>
//#include <boost/python.hpp>
//#include <TMath.h>
//
namespace py = pybind11;

double LnL0(double data, double mc){ // BEWARE this returns -2 * ( log P(n_data|l_model) - log P(n_data|n_data) )

  if(data<0) return 1e12;

  double MCError = 1e-12;
  if(mc > MCError){

    if(data > 0) return 2*(mc - data + data*log(data/mc));
    else         return 2*mc;

  }

  else{

    if(data > MCError) return 2  * (mc - data + data*log(data/MCError)) +
                              data * (3 - 4*mc/MCError + std::pow(mc/MCError,2));

    else if(data > 0)  return mc*mc/MCError + MCError -
                              2*data*(mc/MCError - log(data/MCError));

    else               return mc*mc/MCError + MCError;

  }

}


double LnL(double data, double mc, double s){

  if(data<0) return 1e12;

  if(s<=0) return LnL0(data, mc);

  double MCError = 1e-12;

  double ss = s*s;

  if(mc > MCError){

    if(data > 0){

      double dmc = 1 - ss*mc;

      double sb = 0.5 * (dmc + std::sqrt(4*data*ss + dmc*dmc)) - 1;

      double mcs = mc * (1 + sb);

      return 2*(mcs - data + data*log(data/mcs)) + sb*sb / ss;

    }
    else{
      if(mc*ss < 1) return mc * (2 - mc*ss);
      else          return 1 / ss;
    }

  }
  else{

    if(data > 0){

      if(data != MCError){

        double dmc = 1 - ss*MCError;

        double sb = 0.5 * (dmc + std::sqrt(4*data*ss + dmc*dmc)) - 1;

        double mcs = MCError * (1 + sb);

        double fmc = (mc*mc + data*(MCError - 2*mc)) / MCError;

        return fmc * (data - mcs)/(data - MCError) + 2*data*log(data/mcs) - sb*sb / ss;

      }
      else return std::pow(mc - MCError,2) / (MCError*(1 + MCError*ss));

    }
    else{

      if(MCError*ss < 1) return MCError + mc*mc*(1/MCError - ss);
      else               return 1 / ss;

    }
  }
}

PYBIND11_MODULE(lnl,handle){
    handle.def("LnL", &LnL);
}

//BOOST_PYTHON_FUNCTION_OVERLOADS(lnl_overloads, LnL, 3, 3)

//BOOST_PYTHON_MODULE(lnl){
//      //boost::python::def("LnL", LnL, lnl_overloads());
//      boost::python::def("LnL", LnL);
//}

/*
int main()
{
    double data;
    double mc;
    double s;

    std::cin >> data;
    std::cin >> mc;
    std::cin >> s;

    std::cout<<("%.16f \n", LnL(data, mc, s)) << std::endl;

    return 0;

}
*/


functions.py
import ROOT
import numpy as np
import h5py
import sys
import os
import km3flux
import matplotlib.pyplot as plt
import boost_histogram as bh
import lnl
from plot_functions import plothist, plothist2d


OscProbDir = "/pbs/throng/km3net/software/oscprob/master-root_6.18.04"
osc_prob_lib = OscProbDir + "/libOscProb.so"
ROOT.gSystem.Load(osc_prob_lib)
op = ROOT.OscProb


def _infinite(scalar):  # From km3services/_tools.py
    """Creates an infinite generator from a scalar.

    Useful when zipping finite length iterators with single scalars.
    """
    while True:
        yield scalar


def _zipequalize(*iterables):  # From km3services/_tools.py
    """Creates equal length iterables for a mix of single and n-length arrays"""

    dims = set(map(len, iterables))
    if len(dims) == 1:
        return iterables

    if 1 not in dims or len(dims) > 2:
        raise ValueError("Input arrays dimensions mismatch.")

    out = []
    for it in iterables:
        if len(it) == 1:
            out.append(_infinite(it[0]))
        else:
            out.append(it)

    return out


def _pdgid2flavor(pdgid):
    """Converts PDG ID to OscProb flavor"""

    if abs(pdgid) == 12:
        return 0
    if abs(pdgid) == 14:
        return 1
    if abs(pdgid) == 16:
        return 2
    raise ValueError("Unsupported PDG ID, please use neutrinos")


def _osc_prob(pmns, prem, flav_in, flav_out, energies, cos_zenith):
    params = _zipequalize(flav_in, flav_out, energies, cos_zenith)

    # use a PremModel to make the paths through the earth
    # with the class PremModel
    # chose an angle for the neutrino and fill the paths with cosTheta,
    # e.g. cosTheta = -1 (vertical up-going)
    P = []
    for fl_in, fl_out, E, cosZ in zip(*params):
        if fl_in < 0:
            pmns.SetIsNuBar(True)
        else:
            pmns.SetIsNuBar(False)
        prem.FillPath(cosZ)
        pmns.SetPath(prem.GetNuPath())
        P.append(pmns.Prob(_pdgid2flavor(fl_in), _pdgid2flavor(fl_out), E))
    # print("Here I am")
    # print(P)
    return P


def use_oscprob(flavour_in, flavour_out, energies, cos_zeniths, nu_params=None):
    """
    Use oscprob from Zineb's scripts
    """

    # oscprob config
    pmns = op.PMNS_Fast()
    prem = op.PremModel()

    ORCA_DEPTH_KM = 2.4  # km
    prem.SetDetPos(6371.0 - ORCA_DEPTH_KM)  # values from Zineb

    # nufit parameters with SK atmospheric data. values from Nufit2022 in http://www.nu-fit.org/?q=node/256
    print(nu_params)
    if nu_params is None:
        nu_params = {
            "dm_21": 7.41e-5,
            "dm_31": 2.507e-3,  # if not inv_hierarchy else dm_21 - 2.465e-3
            "theta_12": 33.41 * np.pi / 180,
            "theta_23": 42.2 * np.pi / 180,  # if inverted ordering 49.0 * np.pi/180
            "theta_13": 8.58 * np.pi / 180,  # if inverted ordering 8.57 * np.pi/180
            "dcp": 232 * np.pi / 180,  # if inverted ordering 276 * np.pi / 180
        }

    pmns.SetDm(2, nu_params["dm_21"])  # set delta_m21 in eV^2
    pmns.SetDm(3, nu_params["dm_31"])  # set delta_m31 in eV^2
    pmns.SetAngle(1, 2, nu_params["theta_12"])  # set Theta12 in radians
    pmns.SetAngle(1, 3, nu_params["theta_13"])  # set Theta13 in radians
    pmns.SetAngle(2, 3, nu_params["theta_23"])  # set Theta23 in radians
    pmns.SetDelta(1, 3, nu_params["dcp"])  # set Delta13 in radians
    print(pmns.GetDm(3))
    print(pmns.GetAngle(2, 3))

    p = _osc_prob(
        pmns,
        prem,
        flavour_in,
        flavour_out,
        energies,
        cos_zeniths,
    )

    return p


def compute_osc_weight(nu_type, energy, dir_z, is_cc, oscillation=True, nu_params=None):
    """
    Computes oscillation weight based on event properties

    Parameters
    ----------
    nu_type : array
        The flavors (pdgid) of the particles for which to get the weights.
    energy : array
        The energies of the interactions.
    dir_z : array
        The z-dir, cos(theta), of the interaction.
    is_cc : array
        The description of the current of the interaction: 2 is cc, 3 is nc.
    oscillation : bool
        True for consider oscillations, false for not
    nu_params : dict
        A dict to give specific oscillation parameters. Default are the current best nufit values.

    Returns
    -------
    weight : array
        The oscillation weights for each event.
    """
    honda = km3flux.flux.Honda()

    # make sure these are all arrays (there were problems with data frames)
    nu_type = np.array(nu_type)
    energy = np.array(energy)
    dir_z = np.array(dir_z)
    is_cc = np.array(is_cc)

    nu_dict = {
        12: honda.flux(2014, "Frejus", solar="min", averaged="azimuth")["nue"],
        14: honda.flux(2014, "Frejus", solar="min", averaged="azimuth")["numu"],
        -12: honda.flux(2014, "Frejus", solar="min", averaged="azimuth")["anue"],
        -14: honda.flux(2014, "Frejus", solar="min", averaged="azimuth")["anumu"],
    }

    # nufit parameters with SK atmospheric data. values from Nufit2022 in http://www.nu-fit.org/?q=node/256
    if nu_params is None:
        nu_params = {
            "dm_21": 7.42e-5,
            "dm_31": 2.510e-3,
            "theta_12": 33.45 * np.pi / 180,
            "theta_23": 42.1 * np.pi / 180,
            "theta_13": 8.62 * np.pi / 180,
            "dcp": 230 * np.pi / 180,
        }

    weight = np.zeros(len(nu_type))

    # make sure the correct is_cc convention is used
    # print(is_cc)
    cc_mask = (
        is_cc == 1
    )  # SOMETHING WRONG HERE!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    # print(cc_mask)
    nc_mask = np.invert(cc_mask)
    # print(cc_mask)
    # print(nu_type[cc_mask])

    for flav in [12, 14]:
        print(flav)

        # get oscillation probs for CC events
        if oscillation:
            # convention: use - dir_z for this
            # flavour_in,flavour_out,...
            # print(flav*np.sign(nu_type[cc_mask]))
            # print(nu_type[cc_mask])

            osc_prob_cc = use_oscprob(
                flav * np.sign(nu_type[cc_mask]),
                nu_type[cc_mask],
                energy[cc_mask],
                -dir_z[cc_mask],
                nu_params,
            )
            # print(osc_prob_cc)

        else:
            osc_prob_cc = np.zeros(np.count_nonzero(cc_mask))
            flavor_out_is_same_as_flavor_in_mask = (
                flav * np.sign(nu_type[cc_mask]) == nu_type[cc_mask]
            )
            osc_prob_cc[flavor_out_is_same_as_flavor_in_mask] = np.ones(
                np.count_nonzero(flavor_out_is_same_as_flavor_in_mask)
            )

        # get flux in a loop as it only takes single numbers (as of yet)
        flux = []
        names = np.vectorize(nu_dict.get)(flav * np.sign(nu_type))
        flux = [
            nu_dict[flav * np.sign(nu_type[i])](energy[i], dir_z[i])
            for i in range(len(nu_type))
        ]

        flux = np.asarray(flux)

        # set all the CC weights: osc_prob times flux
        weight[cc_mask] += osc_prob_cc * flux[cc_mask]

        # set all NC weights; total flux, consists of e & mu, no oscillations
        weight[nc_mask] += flux[nc_mask]
        # print(nc_mask)
        # print(cc_mask)

    return weight


def compute_evt_weight(
    nu_type, energy, dir_z, is_cc, w2, livetime, ngen, oscillation=True, nu_params=None
):
    """
    Computes event weight with oscillations

    Parameters
    ----------
    nu_type : array
        The flavors (pdgid) of the particles for which to get the weights.
    energy : array
        The energies of the interactions.
    dir_z : array
        The z-dir, cos(theta), of the interaction.
    is_cc : array
        The description of the current of the interaction: 2 is cc, 3 is nc.
    oscillation : bool
        True for consider oscillations, false for not
    nu_params : dict
        A dict to give specific oscillation parameters. Default are the current best nufit values.

    Returns
    -------
    weight : array
        The weights for each event including the oscillations.
    """

    # make sure these are all arrays (there were problems with data frames)
    nu_type = np.array(nu_type)
    energy = np.array(energy)
    dir_z = np.array(dir_z)
    is_cc = np.array(is_cc)
    w2 = np.array(w2)
    livetime = np.array(livetime)
    ngen = np.array(ngen)

    oscilation_w = compute_osc_weight(
        nu_type, energy, dir_z, is_cc, oscillation, nu_params
    )
    weight = livetime * w2 / ngen * oscilation_w

    return weight



def compute_chi2_map(E, dir_z, data, mc, Ebins=30, dir_zbins=20):
    """
    Computes the chi2 map from a 2d histogram of data and mc values

    Parameters
    ----------
    E : array
        Energies of the events.
    dir_z : array
        Direction of the events as they come from the DSTs.
    data : array
        Weight of each individual data event.
    mc : array
        Weight of each individual mc event.
    Ebins : int
        Number of bins in the E axis.
    dir_zbins : int
        Number of bins in the dir_z axis.


    Returns
    -------
    h : boost_histogram object
        Boost histogram object containing the 2d histogram of dir_z vs E filled with the corresponding Chi2 value for each bin.
    """

    hdata = bh.Histogram(
        bh.axis.Regular(Ebins, 1, 100, transform=bh.axis.transform.log),
        bh.axis.Regular(dir_zbins, -1, 0),
    )
    hmc = bh.Histogram(
        bh.axis.Regular(Ebins, 1, 100, transform=bh.axis.transform.log),
        bh.axis.Regular(dir_zbins, -1, 0),
    )
    hdata2 = bh.Histogram(
        bh.axis.Regular(Ebins, 1, 100, transform=bh.axis.transform.log),
        bh.axis.Regular(dir_zbins, -1, 0),
    )
    hmc2 = bh.Histogram(
        bh.axis.Regular(Ebins, 1, 100, transform=bh.axis.transform.log),
        bh.axis.Regular(dir_zbins, -1, 0),
    )
    hdata.fill(E, -1.0 * dir_z, weight=data)
    hmc.fill(E, -1.0 * dir_z, weight=mc)
    hdata2.fill(E, -1.0 * dir_z, weight=data**2)
    hmc2.fill(E, -1.0 * dir_z, weight=mc**2)
    s = np.sqrt(hmc2.values()) / hmc.values()

    # vectorize LnL function
    vLnL = np.vectorize(lnl.LnL)
    chi2 = vLnL(hdata.values(), hmc.values(), s)

    h = bh.Histogram(
        bh.axis.Regular(Ebins, 1, 100, transform=bh.axis.transform.log),
        bh.axis.Regular(dir_zbins, -1, 0),
    )
    h[...] = np.stack(chi2.T, axis=-1)

    return h


define_binning.py
import boost_histogram as bh
import uproot 
import numpy as np
import matplotlib.pyplot as plt
from scipy.interpolate import interp1d
from plot_functions import *


def cumulative_1d(hist: bh.Histogram, **kwargs):
    """
    Returns a function of the cumulative distribution of a variable based on the provided boost_histogram object.

    Parameters
    ----------
    hist : boost_histogram.Histogram
        A 1D histogram object from the boost_histogram library.

    Returns
    -------
    cdf : callable
        A function that takes a single numerical argument x and returns the cumulative distribution
        of the variable based on the provided histogram object. If the histogram is 2D, the function
        will calculate the marginal cumulative distribution along the x-axis.

    Raises
    ------
    ValueError
        If the input histogram is not 1D .

    Example
    -------
    >>> import boost_histogram as bh
    >>> import numpy as np
    >>> hist = bh.Histogram(bh.axis.Regular(10, 0, 1))
    >>> hist.fill(np.random.rand(1000))
    >>> cdf_func = cumulative_1d(hist)
    >>> cdf_func(0.5)
    0.494
    """
    if hist.ndim not in [1]:
        raise ValueError("Input histogram must be 1D.")
    hcum = np.cumsum(hist.values())
    hcum = hcum/np.amax(hcum)
    hcum = np.insert(hcum, 0, 0.0)
    fcum = interp1d(h.axes.edges[0], hcum, kind = "linear")
    return cdf


    

def define_binning_function(fpath: str, is_pid: bool = False, pid: str = "low_track"):
    # Define the binning for the histograms in
    # Energy and cos theta
    default_theta = 20
    default_E = 30
    lowlim = 1
    uplim = 100
    f = uproot.open(fpath)
    E = f["sel:energy_recoJEnergy"].array().to_numpy()
    if is_pid:
        pid_proba_track = f["sel:pid_proba_track"].array().to_numpy()
        antimu_proba_bkg = f["sel:antimu_proba_bkg"].array().to_numpy()
        classes ={
           "low_track":(E>lowlim) & (E<uplim) & (pid_proba_track > 0.85) & (antimu_proba_bkg <= 1e-4), 
           "low_shower":(E>lowlim) & (E<uplim) & (pid_proba_track <= 0.85) & (antimu_proba_bkg <= 1e-4), 
           "high_track":(E>lowlim) & (E<uplim) & (pid_proba_track > 0.85) & (antimu_proba_bkg > 1e-4) & (antimu_proba_bkg <= 2e-3),
           "high_shower":(E>lowlim) & (E<uplim) & (pid_proba_track <= 0.85) & (antimu_proba_bkg > 1e-4) & (antimu_proba_bkg <= 2e-3),
        }
        condition = classes[pid]
        
    else:
        condition = (E>lowlim) & (E<uplim)
    E = E[condition]
    dir_z = -1.0 * f["sel:cos_zenith_recoJGandalf"].array().to_numpy()[condition]
    #dir_z = dir_z[dir_z>0.0]
    #dir_z = dir_z[dir_z<1.0]
    #E = f["sel:energy_true"].array().to_numpy()
    #dir_z = -1.0 * f["sel:cos_zenith_true"].array().to_numpy()
    w2 = f["sel:w2"].array().to_numpy()/1e17
    w = f["sel:wOsc"].array().to_numpy()[condition]
    ngen = f["sel:ngen"].array().to_numpy()


    hcostheta = bh.Histogram(
        bh.axis.Regular(default_theta, -1, 0)
    )
    hE = bh.Histogram(
        bh.axis.Regular(default_E, lowlim, uplim, transform=bh.axis.transform.log)
    )
    h = bh.Histogram(
        bh.axis.Regular(default_E, lowlim, uplim, transform=bh.axis.transform.log),
        bh.axis.Regular(default_theta, -1, 0),
    )
    hcostheta.fill(-1.0*dir_z, weight=w)
    hE.fill(E, weight=w)
    h.fill(E, -1.0*dir_z, weight=w)
    fcum = cumulative_1d(hcostheta)
    Ecum = np.cumsum(hE.values())
    Ecum = np.insert(Ecum, 0, 0.0)
    fcumE = interp1d(hE.axes.edges[0], Ecum, kind = "linear")
    #plt.plot(hE.axes.edges[0], fcumE(hE.axes.edges[0]))
    #plt.show()
    hcostheta_flat = bh.Histogram(
        bh.axis.Regular(20, 0, np.amax(costhetacum)+0.0001)
    )
    hnew = bh.Histogram(
        bh.axis.Regular(default_E, 1, 100, transform=bh.axis.transform.log),
        #bh.axis.Regular(default_E, 0, np.amax(Ecum)+0.0001),
        bh.axis.Regular(default_theta, 0, np.amax(costhetacum)+0.0001),
    )
    #hcostheta_flat.fill(fcum(-1.0*dir_z), weight=w2*E**(-2)/ngen[0])
    hcostheta_flat.fill(fcum(-1.0*dir_z), weight = w)
    #hnew.fill(fcum(-1.0*dir_z), E, weight = w)
    print(Ecum)
    print(fcumE(E))
    #hnew.fill(fcumE(E), fcum(-1.0*dir_z), weight = w)
    hnew.fill(E, fcum(-1.0*dir_z), weight = w)
    return h, hnew


H, Hnew = define_binning_function("/home/spm/Documents/PhD/KM3NeT/local_work/chi2_tool/chi2_tool/SelectedEvents_mc_nu.root")
mii = None #1
scale = 4
fig, ax = plt.subplots(1,2,figsize = (scale*3,scale*1))
im0 = plothist2d(H,ax=ax[0], vmin = mii)
im1 = plothist2d(Hnew,ax=ax[1], vmin = mii)
ax[0].set_xscale("log")
ax[1].set_xscale("log")
im0.cmap.set_under('white')
im1.cmap.set_under('white')
fig.colorbar(im0, ax = ax[0])
fig.colorbar(im1, ax = ax[1])
ax[0].set_title("Original binning")
ax[1].set_title("New binning")
ax[0].set_xlabel(r"E [GeV]")
#ax[1].set_xlabel(r"CDF(E [GeV])")
ax[1].set_xlabel(r"E [GeV]")
ax[0].set_ylabel(r"cos ($\theta$)")
ax[1].set_ylabel(r"CDF(cos ($\theta$))")
plt.tight_layout()
plt.show()
fig, ax = plt.subplots()
plothist(Hnew[:: bh.sum, :], ax = ax)
plt.show()
fig, ax = plt.subplots()
plothist(Hnew[:, 0], ax = ax)
plt.show()
#for item in ["low_track", "low_shower", "high_track", "high_shower"]:
#    H, Hnew = define_binning_function("/home/spm/Documents/PhD/KM3NeT/local_work/chi2_tool/chi2_tool/SelectedEvents_mc_nu.root",is_pid=True, pid = item)
#    fig, ax = plt.subplots(1,2,figsize = (scale*3,scale*1))
#    im0 = plothist2d(H,ax=ax[0], vmin = mii)
#    im1 = plothist2d(Hnew,ax=ax[1], vmin = mii)
#    ax[0].set_xscale("log")
#    im0.cmap.set_under('white')
#    im1.cmap.set_under('white')
#    fig.colorbar(im0, ax = ax[0])
#    fig.colorbar(im1, ax = ax[1])
#    ax[0].set_title("Original binning")
#    ax[1].set_title("New binning")
#    ax[0].set_xlabel(r"E [GeV]")
#    ax[1].set_xlabel(r"CDF(E [GeV])")
#    ax[0].set_ylabel(r"cos ($\theta$)")
#    ax[1].set_ylabel(r"CDF(cos ($\theta$))")
#    fig.suptitle(item)
#    plt.tight_layout()
#    plt.show()

apply_merged_dst.py
import numpy as np
import uproot
import matplotlib.pyplot as plt
import matplotlib as mpl
import boost_histogram as bh
from functions import *
import lnl


f = uproot.open(
    "/sps/km3net/users/spenamar/oscillations/chi2_tool/results/v7.10/SelectedEvents_mc_nu.root"
)  # merged DST of events passing selection
# E = f['sel:energy_true'].array().to_numpy() # True energy
E = f["sel:energy_recoJEnergy"].array().to_numpy()
# dir_z = -1.*f['sel:cos_zenith_true'].array().to_numpy() # True cos_zenith
dir_z = -1.0 * f["sel:cos_zenith_recoJGandalf"].array().to_numpy()
nu_type = f["sel:type"].array().to_numpy()
is_cc = f["sel:is_cc"].array().to_numpy()
print(is_cc)
w2 = f["sel:w2"].array().to_numpy()
livetime = f["sel:run_duration"].array().to_numpy()
ngen = f["sel:ngen"].array().to_numpy()


osci_w = compute_evt_weight(nu_type, E, dir_z, is_cc, w2, livetime, ngen)
nu_params = {
    "dm_21": 7.42e-5,
    "dm_31": 2.210e-3,
    # "dm_31": 0,
    "theta_12": 40.0 * np.pi / 180,
    "theta_23": 70.0 * np.pi / 180,
    # "theta_23": 0. * np.pi/180,
    "theta_13": 8.62 * np.pi / 180,
    "dcp": 230 * np.pi / 180,
}
osci_w_bad = compute_evt_weight(
    nu_type, E, dir_z, is_cc, w2, livetime, ngen, nu_params=nu_params
)
print(osci_w)
print(osci_w_bad)


# plothist = lambda h: plt.bar(*h.axes.centers, h.values(), width=h.axes.widths[0], alpha=0.6);

h1 = bh.Histogram(
    bh.axis.Regular(50, 1, 100, transform=bh.axis.transform.log),
)
h2 = bh.Histogram(
    bh.axis.Regular(50, 1, 100, transform=bh.axis.transform.log),
)

h1.fill(E, weight=osci_w)
h2.fill(E, weight=osci_w_bad)
fig, ax = plt.subplots()
plothist(h1, ax)
plothist(h2, ax)
plt.xscale("log")
plt.show()


Ebins = 30
zbins = 20

h1 = bh.Histogram(
    bh.axis.Regular(Ebins, 1, 100, transform=bh.axis.transform.log),
    bh.axis.Regular(zbins, -1, 0),
)
h2 = bh.Histogram(
    bh.axis.Regular(Ebins, 1, 100, transform=bh.axis.transform.log),
    bh.axis.Regular(zbins, -1, 0),
)
h11 = bh.Histogram(
    bh.axis.Regular(Ebins, 1, 100, transform=bh.axis.transform.log),
    bh.axis.Regular(zbins, -1, 0),
)
h22 = bh.Histogram(
    bh.axis.Regular(Ebins, 1, 100, transform=bh.axis.transform.log),
    bh.axis.Regular(zbins, -1, 0),
)
h1.fill(E, -1.0 * dir_z, weight=osci_w)
h2.fill(E, -1.0 * dir_z, weight=osci_w_bad)
h11.fill(E, -1.0 * dir_z, weight=osci_w**2)
h22.fill(E, -1.0 * dir_z, weight=osci_w_bad**2)
s = np.sqrt(h22.values()) / h2.values()
fig, ax = plt.subplots()
plothist2d(h1, ax)
# plothist(h2)
plt.xscale("log")
plt.show()
fig, ax = plt.subplots()
plothist2d(h2, ax)
plt.xscale("log")
plt.show()


print("#" * 20)
print("#" * 20)
print("#" * 20)
print("#" * 20)
vLnL = np.vectorize(lnl.LnL)
print("h1")
print(h1.values())
print(np.shape(h1.values()))
print("h2")
print(h2.values())
print(np.shape(h2.values()))
print("s")
print(s)
print(np.shape(s))
chi2 = vLnL(h1.values(), h2.values(), s)
print("chi2")
print(chi2)
print(np.shape(chi2))
print("#" * 20)
print("#" * 20)
print("#" * 20)
print("#" * 20)

fig, ax = plt.subplots()
h = compute_chi2_map(E, dir_z, osci_w, osci_w_bad)

plothist2d(h, ax)
plt.xscale("log")
plt.show()
